### DISCUSSION TITLE
[@types/sinon] missing TType union on SinonStubbedInstance

### DISCUSSION THREAD

**Discussion by azatoth** [DefinitelyTyped/DefinitelyTyped#56694] [Issues with a @types package] [answered]
at the moment we have the following type definition

```typescript
   type SinonStubbedInstance<TType> = {
        [P in keyof TType]: SinonStubbedMember<TType[P]>;
    };
```

This prevents this object to be used directly as the original type and ugly typecastings (like `run(fooStubInstance as any)`) have to be made.

I recommend adding `TType` as a union to the type:
```typescript
    type SinonStubbedInstance<TType> = {
        [P in keyof TType]: SinonStubbedMember<TType[P]>;
    } & TType;
```

[a comment by an automated account is not shown]

**Comment by 43081j** [comment, marked as answer]
im wondering if that'll work for every case. there's no chance whats in `TType` could override stub members? typescript will merge what it can but just concerned there may be some types that don't merge, and instead overwrite. could be wrong though

**Reply by azatoth** [reply]
> im wondering if that'll work for every case. there's no chance whats in `TType` could override stub members? typescript will merge what it can but just concerned there may be some types that don't merge, and instead overwrite. could be wrong though

Would switching the order to `TType & {...}` alleviate this? The important part is to tell typescript this object is `TType` compatible.

**Reply by 43081j** [reply]
that would work. though typescript will check the type by its shape anyway, so im surprised a stubbed instance doesn't work for you.

could you possibly make a minimum example of where the current type doesn't work?

**Reply by azatoth** [reply]
I assume it might be mostly due to non-public methods and fields;

```typescript
import chai, { expect } from 'chai';

import sinon from 'sinon';
import sinonChai from 'sinon-chai';

chai.use(sinonChai);

class Foo {
  private myVar: string;

  constructor() {
    this.myVar = '42';
  }

  public getMyVar() {
    return this.myVar;
  }
}

function doit(instance: Foo) {
  return instance.getMyVar();
}

describe('my test', () => {
  let fooInstance: sinon.SinonStubbedInstance<Foo>;

  beforeEach(() => {
    fooInstance = sinon.createStubInstance(Foo);
  });

  it('should work', () => {
    fooInstance.getMyMVar.returns('84');
    expect(doit(fooInstance)).to.eql('84');
  });
});

```
Triggers following error:
```
Argument of type 'SinonStubbedInstance<Foo>' is not assignable to parameter of type 'Foo'.
  Property 'myVar' is missing in type 'SinonStubbedInstance<Foo>' but required in type 'Foo'.
```

**Reply by 43081j** [reply]
yep that makes sense. i think your suggestion seems ok then 👍 

feel free to open a PR, or i can at some point if you don't think you'll get chance. would be good to test it a bit more, just be sure its the right change to make and won't break anything

**Reply by azatoth** [reply]
> yep that makes sense. i think your suggestion seems ok then +1
> 
> feel free to open a PR, or i can at some point if you don't think you'll get chance. would be good to test it a bit more, just be sure its the right change to make and won't break anything

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56741 created
